packages/reakit/src/Form/FormState.ts

```diff
@@ -306,23 +303,29 @@ export function unstable_useFormState<V = Record<any, any>>(
   });
 
   const validate = React.useCallback(
-    async (vals = state.values) => {
-      try {
-        if (onValidate) {
-          const response = onValidate(filterAllEmpty(vals));
-          if (isPromise(response)) {
-            dispatch({ type: "startValidate" });
+    (vals = state.values) =>
+      Promise.resolve()
+        .then(() => {
+          if (onValidate) {
+            const response = onValidate(filterAllEmpty(vals));
+            if (isPromise(response)) {
+              dispatch({ type: "startValidate" });
+            }
+            return response;
           }
-          const messages = await response;
-          dispatch({ type: "endValidate", messages });
+          return undefined;
+        })
+        .then(messages => {
+          if (messages) {
+            dispatch({ type: "endValidate", messages });
+          }
```

Comment:
It's possible that the `onValidate` function returns without messages. So I guess the check should be removed.

Reply:
It couldn't be removed because otherwise it would run not only after a `onValidate` call but also after the `return undefined`. I refactored it.